File: 2023/go/pkg/day4/problem2.go
package day4

import (
	"fmt"
	"slices"
	"strconv"
	"strings"
)

func SolveProblem2(lines []string) string {
	sum := 0

	for _, line := range lines {
		newCount := processCard(line, lines)

		for _, v := range newCount {
			sum += v
		}
	}

	return fmt.Sprint(sum)
}

func processCard(card string, lines []string) map[string]int {
	cards := strings.Split(card, ": ")
	id, _ := strconv.Atoi(strings.Fields(cards[0])[1])
	allnums := strings.Split(cards[1], " | ")

	cardCount := map[string]int{
		strconv.Itoa(id): 1,
	}

	winning := strings.Fields(allnums[0])
	mynums := strings.Fields(allnums[1])

	count := 0

	for _, num := range mynums {
		if slices.Contains(winning, num) {
			count++
		}
	}

	for i := id; i < count+id; i++ {
		newCount := processCard(lines[i], lines)

		for k, v := range newCount {
			cardCount[k] += v
		}
	}

	return cardCount
}

File: 2023/rust/src/day4/problem2.rs
pub fn solve(lines: Vec<&str>) -> String {
    let mut sum = 0;

    for line in lines.iter() {
        sum += process_card(line, &lines);
    }

    sum.to_string()
}

fn process_card(line: &str, lines: &Vec<&str>) -> i32 {
    let card: Vec<&str> = line.split(": ").collect();
    let id: i32 = card[0].split_whitespace().last().unwrap().parse().unwrap();
    let allnums: Vec<&str> = card[1].split(" | ").collect();
    let winning: Vec<&str> = allnums[0].split_whitespace().collect();

    let mut sum = 1;
    let mut count = id as usize;

    for num in allnums[1].split_whitespace() {
        if winning.contains(&num) {
            sum += process_card(lines[count], lines);
            count += 1;
        }
    }

    sum
}
